zaza/openstack/charm_tests/ceph/tests.py

```diff
@@ -690,28 +701,201 @@ def expected_apps(self):
+
+        :param app_name: Secondary site rgw Application to be promoted.
+        :type app_name: str
+        """
+        if app_name is self.primary_rgw_app:
+            new_secondary = self.secondary_rgw_unit
+        else:
+            new_secondary = self.primary_rgw_unit
+
+        # Promote to Primary
+        zaza_model.run_action_on_leader(
+            app_name,
+            'promote',
+            action_params={},
+        )
+
+        # Period Update Commit new secondary.
+        cmd = self.get_rgwadmin_cmd_skeleton(new_secondary)
+        zaza_model.run_on_unit(
+            new_secondary, cmd + 'period update --commit'
+        ).get('Stdout', '')
+
+    def get_client_keys(self, rgw_app_name=None):
+        """
+        Create access_key and secret_key for boto3 client.
+
+        :param rgw_app_name: RGW application for which keys are required.
+        :type rgw_app_name: str
+        """
+        unit_name = self.primary_rgw_unit
+        if rgw_app_name is not None:
+            unit_name = rgw_app_name + '/0'
+        user_name = 'botoclient'
+        cmd = self.get_rgwadmin_cmd_skeleton(unit_name)
+        users = json.loads(zaza_model.run_on_unit(
+            unit_name, cmd + 'user list'
+        ).get('Stdout', ''))
+        # Fetch boto3 user keys if user exists.
+        if user_name in users:
+            output = json.loads(zaza_model.run_on_unit(
+                unit_name, cmd + 'user info --uid={}'.format(user_name)
+            ).get('Stdout', ''))
+            keys = output['keys'][0]
+            return keys['access_key'], keys['secret_key']
+        # Create boto3 user if it does not exist.
+        create_cmd = cmd + 'user create --uid={} --display-name={}'.format(
+            user_name, user_name
+        )
+        output = json.loads(
+            zaza_model.run_on_unit(unit_name, create_cmd).get('Stdout', '')
+        )
+        keys = output['keys'][0]
+        return keys['access_key'], keys['secret_key']
+
+    @tenacity.retry(
+        retry=tenacity.retry_if_result(lambda ret: ret is None),
+        wait=tenacity.wait_fixed(10),
+        stop=tenacity.stop_after_attempt(5)
+    )
+    def get_rgw_endpoint(self, unit_name: str):
+        """
+        Fetch Application endpoint for RGW unit.
+
+        :param unit_name: Unit name for which RGW endpoint is required.
+        :type unit_name: str
+        """
+        unit = zaza_model.get_unit_from_name(unit_name)
+        unit_address = zaza_model.get_unit_public_address(
+            unit,
+            self.model_name
+        )
+
+        logging.debug("Unit: {}, Endpoint: {}".format(unit_name, unit_address))
+        if unit_address is None:
+            return None
+        # Evaluate port
+        try:
+            zaza_model.get_application("vault")
+            return "https://{}:443".format(unit_address)
+        except KeyError:
+            return "http://{}:80".format(unit_address)
+
+    def configure_rgw_apps_for_multisite(self):
```

Comment:
Well, The app will settle once a. Config is done (being set by this method), b. relation is added, c. The migration magic is done. Since this is only a single piece from the total process (just configures the multisite values), we can't wait "just yet" for the app to settle.

Reply:
That's not what I meant. The function may return in terms of the sync wrapper that is `set_application_config`, but that doesn't mean that the configuration has been actually applied to the model. For that reason, I recommend inserting calls to `wait_for_application_states` whenever an "important" part of the configuration is modified, or when a long-running action is executed.

Reply:
Oh okay, I see what you meant.
We use this method twice and in both scenarios the overall expectation to wait for is different:
1. In the secondary block scenario we expect the secondary app to go to blocked state with certain message.
2. In Normal migration scenario we expect it to block and then subsequently be ready.

For this mix/match of expectations I have kept the respective waiting logic in the actual test. Usage in test_004 and test_005.